### ISSUE TITLE
Code just doesn't run no idea why ...

### ISSUE THREAD

**Issue by TehWardy** [dotnet/roslyn#42562]
**Version Used**: 
 Latest (pulled yesterday from github, 3,5,0 I think)
**Steps to Reproduce**:
 I have this method in my code ...
```cs
public async Task<T> Run<T>(string code, string[] imports, object args, Action<WorkflowLogLevel, string> log)
{
    try
    {
        var referencesNeeded = references.Where(r => r.GetExportedTypes().Any(t => imports.Contains(t.Namespace)));
        var options = ScriptOptions.Default
            .AddReferences(referencesNeeded)
            .WithImports(imports);

        if (log != null)
        {
            var message = $"\nImports\n  {string.Join("\n  ", imports)}\n\nReferences Needed\n  {string.Join("\n  ", referencesNeeded.Select(r => r.FullName))}";
            log(WorkflowLogLevel.Debug, message);
        }

        return await CSharpScript.EvaluateAsync<T>(code, options, args, args.GetType());
    }
    catch (NullReferenceException ex)
    {
        var typeAndCall = $"(({ex.TargetSite.DeclaringType.Name})object).{ex.TargetSite.Name}";
        var data = new List<string>();

        foreach (var k in ex.Data.Keys) data.Add($"{k}: {ex.Data[k]}");

        throw new Exception(ex.Message + $"\nContext: {ex.Source}\nTarget: {typeAndCall}\n{string.Join("\n", data)}");
    }
    catch (CompilationErrorException ex)
    {
        throw new Exception($"Compilation failed:\n{ex.Message}\n{string.Join(Environment.NewLine, ex.Diagnostics)}");
    }
}
```

**Expected Behavior**:
When executed it should run the code given to it.

**Actual Behavior**:
It hits the logging call just before the roslyn call and logs out the dependencies fine, then nothing!
My code appears to just stop, the whole application is just frozen there.

The code given to it looks like this (in case it matters) ...

```cs
Func<string, string, string> MapSystem = (sys, defaultVal) => {
	switch(sys)
	{
		case "VATRef": return "GlobalTax";
		default: return defaultVal;
	}
};

if (flow.GetActivity<DMSCreateTextFilesActivity>("CopyMD2DMS").Contents != null){
((ProposeCompanies)activity).Proposals = flow.GetActivity<DMSCreateTextFilesActivity>("CopyMD2DMS").Contents
	.Select(i => Data.ParseXml<dynamic>(i))
	.SelectMany((dynamic f) => ((JArray)f.Companies.Company)
		.Select((dynamic c) => new ProposedCompany {
			Name = c.Name,
			//ContactEmailAddress
			//ContactPhoneNumber
			//WebsiteUrl
			//DateTradingEnabled
			Address = new Address {
				CountryId = c.Address.CountryCode,
				//PoBox 
				Line1 = c.Address.Address1,
				Line2 = c.Address.Address2,
				ZipOrPostalCode = c.Address.PostCode,
				TownOrCity = c.Address.City,
				StateOrProvince = c.Address.Region
			},
			References = (c.References.Reference is JArray) 
				? ((JArray)c.References.Reference)
					.Select((dynamic r) => new ProposedCompanyReference {
						SystemId = MapSystem(((IDictionary<string, JToken>)r).Keys.First().ToString(), f.Companies.SourceSystem.ToString()),
						Value = ((IDictionary<string, JToken>)r).First().Value.ToString()
					}).ToArray()
				: new [] { 
					new ProposedCompanyReference {
						SystemId = MapSystem(((IDictionary<string, JToken>)c.References.Reference).Keys.First().ToString(), f.Companies.SourceSystem.ToString()),
						Value = ((IDictionary<string, JToken>)c.References.Reference).First().Value.ToString()
					}
				},
			Buckets = new [] { new BucketProposedCompany { BucketId = new Guid(variables["MasterDataBucket"].ToString()) } }
		})
	)
	.ToArray();
};
```

**Comment by TehWardy** [comment]
In all honesty I think this is a problem with our code rather than roslyn, but my understanding is that I shouldn't be in a situation where roslyn acts like a black hole and never returns right?

**Comment by tmat** [comment]
@TehWardy I'd recommend adding some logging to the script you're executing to find out where it gets stuck.

**Comment by TehWardy** [comment]
sorry there's already logging in it, the first and last lines I cut out as I didn't think they were relevant here, it doesn't appear to be hitting the first logging call though which is why i suspected something about the compiler being at fault.

**Comment by TehWardy** [comment]
Found this in the systems event log when the crash happens ...
```
Faulting application name: WorkflowConsole.exe, version: 1.0.0.0, time stamp: 0x5dedc13b
Faulting module name: coreclr.dll, version: 4.700.19.60701, time stamp: 0x5debd517
Exception code: 0xc00000fd
Fault offset: 0x0000000000140994
Faulting process id: 0x18f4
Faulting application start time: 0x01d6067aa0e0de88
Faulting application path: D:\Workflow\WorkflowConsole.exe
Faulting module path: D:\Workflow\coreclr.dll
Report Id: 42e4480e-bc33-4331-9880-c99ce602ca83
Faulting package full name: 
Faulting package-relative application ID: 
```

**Comment by TehWardy** [comment]
Ok i managed to pin it down to this code ...
```cs
var stuff = copyStep.Contents.Select(i => Core.Objects.Data.ParseXml<dynamic>(i));
```
ParseXml looks like this ...

```cs
public static T ParseXml<T>(string data)
{
    var builder = new StringBuilder();
    JsonSerializer.Create().Serialize(new CleanJsonWriter(new StringWriter(builder)), ParseXml(data));
    return JsonConvert.DeserializeObject<T>(builder.ToString());
}
```
and depends on this ...

```cs
internal class CleanJsonWriter : JsonTextWriter
{
    public CleanJsonWriter(TextWriter writer) : base(writer) { }

    public override void WritePropertyName(string name)
    {
        var result = name;
        if (name.StartsWith("@") || name.StartsWith("#"))
            result = name.Substring(1);

        if (result.Contains(":"))
            result = result.Split(":".ToCharArray()).Last();

        base.WritePropertyName(result);
    }
}
```
There's probably a better way of achieving this result but this is mostly existing code in existing frameworks that proven to work that i'm stitching together so I figured it'll do for now until I need to optimise it further.

As to why it causes a crash I have no idea.

**Comment by TehWardy** [comment]
Ok i managed to pin it down to dynamic operations where it can't resolve the type info for some reason.
The issue is definitely in my code (which I have now managed to fix), but I think it should at least catch the resolution failure and throw some sort of exception.

I've also seen it happen in here ...
```cs
    public static class JObjectExtensions
    {
        // for use with dynamic selections where the parsed results "might" be an object or might be an array
        public static TResult[] JObjectOrJArray<T, TResult>(this object source, Func<T, TResult> select) => source switch
        {
            JArray   a => a.JObjectOrJArray(select),
            IDynamicMetaObjectProvider jo => jo.JObjectOrJArray(select),
            _ => new TResult[0]
        };

        public static TResult[] JObjectOrJArray<T, TResult>(this JArray source, Func<T, TResult> select)  => source.Cast<T>().Select(select).ToArray();

        public static TResult[] JObjectOrJArray<T, TResult>(this IDynamicMetaObjectProvider source, Func<T, TResult> select) where T : IDynamicMetaObjectProvider => new[] { select((T)source) };
    }
```

... so i refactored it to this ...

```cs
public static class JObjectExtensions
{
    // for use with dynamic selections where the parsed results "might" be an object or might be an array
    public static TResult[] JObjectOrJArray<T, TResult>(this object source, Func<T, TResult> select) => source switch
    {
        JArray   a => a.JArray(select),
        JObject jo => jo.JObject(select),
        _ => new TResult[0]
    };

    static TResult[] JArray<T, TResult>(this JArray source, Func<T, TResult> select)  => source.Cast<T>().Select(select).ToArray();

    static TResult[] JObject<T, TResult>(this JObject source, Func<T, TResult> select) => new[] { source }.Cast<T>().Select(select).ToArray();
}
```

usage for this is something like ...

```cs
var jsonString = "{  ... }";
var results = JsonConvert.DeserializeObject(jsonString).Foos.JObjectOrJArray<dynamic, TResult>(f => ...);
```

I built this because i'm working with XML without schema's so sometimes when an array has only one child the .Net parser makes it a single object using the parsing code above and this reuslts in a single JObject instead of a JArray after the JsonConversion.

The idea is that the system can take in any valid XML and then return a JObject for that XML Document which it can then treat like a dynamic object and run user defined code on as part of a workflow.

Long story short, it's complex as hell, and I hate having to use dynamic like this but the alternative is to build a ton of tooling I don't yet have so this is the interim solution.
My code is often "flaky" but I've built awesome logging and reporting around it to handle weird exceptions, I just need Roslyn to tell me what they are.

**Comment by TehWardy** [comment]
Further details show that it only seems to occur when i have a "large" set of data.
Assuming the input data is about 2MB of XML the above ParseXml method when called as part of a dynamic operation dispatched from my code that passes in the XML string causes this behaviour.

It seems to be some sort of data size limit for Roslyn scripts.
Is there a way to overcome this?

I would like to set a limit of at least 100MB.

**Comment by CyrusNajmabadi** [comment]
Closing out as stale.
